File: catalog/backends/ruby-rails/patterns/audit_log_middleware.rb
# Pattern: SIEM — structured audit log for auth-relevant events
# Adapt: log fields relevant to your domain's security events
require "json"

class AuditLogMiddleware
  def initialize(app)
    @app = app
    @logger = Logger.new($stdout)
  end

  def call(env)
    status, headers, response = @app.call(env)
    @logger.info({
      method: env["REQUEST_METHOD"],
      path: env["PATH_INFO"],
      status: status,
      remote: env["REMOTE_ADDR"]
    }.to_json)
    [status, headers, response]
  end
end

File: catalog/backends/go-gin/patterns/audit_log_middleware.go
package patterns

import (
	"log/slog"
	"time"

	"github.com/gin-gonic/gin"
)

// Pattern: SIEM — structured audit log for every request.
// Adapt: add authentication fields (user ID, role) from your auth layer.
func AuditLogMiddleware() gin.HandlerFunc {
	return func(c *gin.Context) {
		start := time.Now()
		c.Next()
		slog.Info("audit",
			"method", c.Request.Method,
			"path", c.Request.URL.Path,
			"status", c.Writer.Status(),
			"client_ip", c.ClientIP(),
			"user_agent", c.Request.UserAgent(),
			"latency_ms", time.Since(start).Milliseconds(),
		)
	}
}
